In this notebook, we are going to use RiderLog dataset, which contains all bicycle trips collected by RiderLog between 2010 and 2013 in Greater Sydney region, thanks to [CityData](https://citydata.be.unsw.edu.au/layers/geonode%3AGreaterSydney_NSW_RiderLog_2010to2013).<br \><br \>

The original data is in a point format, where bicycle trips are divided into points which are centroids of cells in a 500m * 500m grid. In order to better explore the relationship between attributes of routes, we kept only one records for each RouteId. While sacrificing some of the spatial information, the dataset is now more easy for you to play with!<br \><br \>

Let's get started!<br \>

__First, lets import pandas library and the RiderLog data. __<br \>

__point_data__ contains the original dataset, and the duplicates have been droped for you and stored in the variable __route_data__.

In [17]:
import ____ as pd

point_data = pd.read_csv("GreaterSydney_NSW_RiderLog_2010to2013.csv")

route_data = point_data.drop_duplicates(subset='RouteId')

__Appitizer - explore the dataset with _describe_ method:__ <br \>
How many routes are there in the dataset?

In [20]:
route_data.____

(14313, 25)

There should be 14313 routes with 25 attributes attached to each one of them. <br \> <br \>
What is the mean, median and max value for distance? <br \>
Have you observed anything abnormal in the dataset? (_Hint: check the minimum values_)

In [36]:
route_data.describe()

RouteId       RiderID      TopSpeed      Distance      Duration  \
count   14313.000000  14313.000000  14313.000000  14313.000000  14313.000000   
mean    68507.194508   4275.132118     11.178481     11.514460     42.831100   
std     41923.387642   2566.966248      3.779017     10.787518     57.523297   
min       338.000000     95.000000      0.000000      0.220000  -4583.170000   
25%     32556.000000   1909.000000      9.100000      5.540000     22.450000   
50%     65595.000000   4437.000000     11.200000      8.850000     33.720000   
75%    102557.000000   6077.000000     13.300000     13.710000     50.550000   
max    152980.000000   9475.000000     93.800000    176.450000   1612.060000   

            AvSpeed     SA1_start       SA1_end           YOB           Age  \
count  14313.000000  1.431300e+04  1.431300e+04  11134.000000  11134.000000   
mean       4.583713  1.191439e+10  1.188042e+10   1971.192923     42.807077   
std        1.802794  1.375774e+09  7.028357e+08     11.731699     11.731699   
min       -0.020000 -1.000000e+00 -1.000000e+00   1930.000000      4.000000   
25%        3.580000  1.170313e+10  1.170313e+10   1962.000000     35.000000   
50%        4.580000  1.190214e+10  1.180214e+10   1972.000000     42.000000   
75%        5.550000  1.210414e+10  1.210414e+10   1979.000000     52.000000   
max       31.880000  1.200000e+11  1.280215e+10   2010.000000     84.000000   

          startyear    startmonth     startdate      startday     starthour  \
count  14313.000000  14313.000000  14313.000000  14313.000000  14313.000000   
mean    2011.272689      7.089289     15.784881      3.934815     11.928666   
std        1.018989      3.450155      8.615184      1.783099      4.865348   
min     2010.000000      1.000000      1.000000      1.000000      0.000000   
25%     2010.000000      4.000000      8.000000      2.000000      7.000000   
50%     2011.000000      8.000000     16.000000      4.000000     11.000000   
75%     2012.000000     10.000000     23.000000      5.000000     17.000000   
max     2013.000000     12.000000     31.000000      7.000000     23.000000   

           startmin      Cell_SA1      Cell_SA2       Cell_ID  
count  14313.000000  1.431300e+04  1.431300e+04  14313.000000  
mean      29.668833  1.187978e+10  1.187978e+08  18073.520855  
std       17.217581  4.878072e+08  4.878072e+06   5656.056532  
min        0.000000  1.020110e+10  1.020110e+08      0.000000  
25%       15.000000  1.170313e+10  1.170313e+08  15822.000000  
50%       30.000000  1.190214e+10  1.190214e+08  16919.000000  
75%       44.000000  1.200314e+10  1.200314e+08  18724.000000  
max       59.000000  1.280215e+10  1.280215e+08  47197.000000

The __mean, median and max__ value for distance should be __11.51, 8.85, 176.45__, respectively. <br \> <br \>

How many numerical attributes are there in the dataset? Try get this information by getting the shape of the dataframe generated above by _describe_ method.

In [6]:
route_data.describe()._____

(8, 19)

How many of them are integer attributes? <br />

_Hint_: Apart from describing the dataset, Python also knows how to count unique values! See [value_counts()](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.Series.value_counts.html) method.

In [8]:
route_data.dtypes._______

int64      13
object      6
float64     6
dtype: int64

There are __19 numerical attributes__ in the dataset, where __13__ of them contain integer only, and __6__ of them contain floats.<br \><br \>Now let's see if we can findout relationships between attributes! <br \>
__How many records have a Duration larger than 60 mins?__<br \>
Remember to tryout both the __shape__ method and the new __value counting__method!

In [10]:
route_data[route_data['Duration'] ____].shape

(2567, 25)

In [22]:
(route_data['Duration'] > 60).value_counts()

False    11746
True      2567
Name: Duration, dtype: int64

__2567__ out of 14313 routes have a duration longer than 60 mins! <br \><br \>

__For all records matches the descrition above, what is the minimum distance of them?__ <br \> 

Hint: To get information on a single column, you can __index that specific column__ and perform something similar that you have used in the lecturer. Yes! I'm talking about max function. If you get stucked, try this [link](https://docs.python.org/3/library/functions.html)!

In [23]:
route_data[route_data['Duration'] > 60].describe()

RouteId      RiderID     TopSpeed     Distance     Duration  \
count    2567.000000  2567.000000  2567.000000  2567.000000  2567.000000   
mean    68515.951695  4381.999610    12.684651    25.693744   101.648457   
std     41516.347663  2614.426555     3.858263    17.257758    71.539383   
min       378.000000   102.000000     0.000000     0.220000    60.050000   
25%     33682.000000  1922.000000    10.600000    14.220000    68.175000   
50%     65843.000000  4462.000000    12.800000    22.570000    81.610000   
75%    103943.500000  6448.000000    15.000000    31.780000   111.955000   
max    147004.000000  9313.000000    39.300000   176.450000  1612.060000   

           AvSpeed     SA1_start       SA1_end          YOB          Age  \
count  2567.000000  2.567000e+03  2.567000e+03  1978.000000  1978.000000   
mean      4.427293  1.209263e+10  1.191093e+10  1967.555106    46.444894   
std       2.094689  3.039491e+09  1.255301e+09    12.009954    12.009954   
min       0.010000 -1.000000e+00 -1.000000e+00  1930.000000     4.000000   
25%       3.005000  1.170313e+10  1.170313e+10  1960.000000    39.000000   
50%       4.530000  1.210114e+10  1.210114e+10  1968.000000    46.000000   
75%       5.815000  1.240315e+10  1.240315e+10  1975.000000    54.000000   
max      23.050000  1.200000e+11  1.280215e+10  2010.000000    84.000000   

        startyear   startmonth    startdate     startday    starthour  \
count  2567.00000  2567.000000  2567.000000  2567.000000  2567.000000   
mean   2011.25672     7.232178    16.030386     3.887028    10.448773   
std       0.98781     3.413765     8.745462     2.179158     4.467210   
min    2010.00000     1.000000     1.000000     1.000000     1.000000   
25%    2010.00000     4.000000     8.000000     2.000000     7.000000   
50%    2011.00000     8.000000    16.000000     4.000000     9.000000   
75%    2012.00000    10.000000    24.000000     6.000000    15.000000   
max    2013.00000    12.000000    31.000000     7.000000    23.000000   

          startmin      Cell_SA1      Cell_SA2       Cell_ID  
count  2567.000000  2.567000e+03  2.567000e+03   2567.000000  
mean     29.739774  1.200751e+10  1.200751e+08  18179.563303  
std      17.225298  5.517579e+08  5.517579e+06   6539.854857  
min       0.000000  1.020110e+10  1.020110e+08      0.000000  
25%      15.000000  1.170313e+10  1.170313e+08  15601.000000  
50%      30.000000  1.200214e+10  1.200214e+08  17741.000000  
75%      44.000000  1.240115e+10  1.240115e+08  21393.000000  
max      59.000000  1.280215e+10  1.280215e+08  47197.000000

In [25]:
min_distance_long_duration = ___(route_data[route_data['Duration'] > 60]['Distance'])
print('Minimum distance for records which have a duration larger than 60 mins is', min_distance_long_duration)

Minimum distance for records which have a duration larger than 60 mins is 0.22


The minimum distance is __0.22km__ for a trip lasts longer than 60 mins. Emm...

__How many recreational trips have a duration longer than 60 minutes?__ <br \> 
To answer this question, we can first get a subset of the dataset meets the description. <br \>
Then, we can focus on the subset to find out the answer of this question. <br \>
_Hint_:Attribute _RidePurpos_ denotes the purpose of the trip, and the key for recreational trips is _Recreation_.


In [30]:
long_duration_routes = route_data[route_data['Duration'] > 60]

long_duration_routes[long_duration_routes['_____'] == ________].shape

(1412, 25)

There are __1412__ recreational trips with a duration longer than 60 minutes!

__Exercise Question:__
1. What is the top speed for recreational trips with a female rider?
2. What is the most popular destination for recreational trips? 
3. Fasted Top Speed is generated by who?
4. Which rider generated most trips?